examples/iris-classification/Main.hs

```diff
@@ -0,0 +1,119 @@
+    l2 :: Linear
+    } deriving (Generic, Show)
+
+instance Parameterized MLP
+instance Randomizable MLPSpec MLP where
+    sample MLPSpec {..} = MLP 
+        <$> sample (LinearSpec inputFeatures hiddenFeatures0)
+        <*> sample (LinearSpec hiddenFeatures0 hiddenFeatures1)
+        <*> sample (LinearSpec hiddenFeatures1 outputFeatures)
+
+mlp :: MLP -> Tensor -> Tensor
+mlp MLP{..} input = 
+    logSoftmax (Dim 1)
+    . linear l2
+    . relu
+    . linear l1
+    . relu
+    . linear l0
+    $ input
+
+data IrisClass = Setosa | Versicolor | Virginica deriving (Eq, Show, Enum, Generic)
+
+instance FromField IrisClass where
+  parseField b = case b of
+    "Iris-setosa"      -> pure Setosa
+    "Iris-versicolor" -> pure Versicolor
+    "Iris-virginica"   -> pure Virginica
+    _                  -> mzero
+data Iris = Iris { sepalLength :: Float
+                 , sepalWidth  :: Float
+                 , petalLength :: Float
+                 , petalWidth  :: Float
+                 , irisClass   :: IrisClass
+                 } deriving (Generic, Show, FromRecord, FromNamedRecord)
+
+-- without this explicit type signature it can't find the right overlapping instance!
+toTensors :: [Iris] -> (Tensor, Tensor)
+toTensors iris = do
+  -- Iris{..} <- iris 
+  -- let features = [sepalLength , sepalWidth , petalLength , petalWidth]
+  --     classes = fromEnum irisClass 
+  -- pure (asTensor features, asTensor classes)
+
+  -- yucky
+  (asTensor $ getFeatures iris, asTensor $ getClasses iris)
+  where 
+        getFeatures = fmap (\x -> [sepalLength x, sepalWidth x, petalLength x, petalWidth x])
+        getClasses = fmap (\x -> fromEnum $ irisClass x)
+
+-- trainLoop :: Optimizer o => model -> o -> Producer [Iris] m  () -> m model
+trainLoop model optimizer = P.foldM step init done
+  -- where step :: MonadIO m => MLP -> (Tensor, Tensor) -> m MLP
+  where 
+        step model (input, label ) = do
+        -- step model bad = do
+          -- let (input, label) = Main.toTensors bad
+          let loss = nllLoss' label $ mlp model input
+          -- when (iter `mod` 50 == 0) $ do
+          liftIO $ putStrLn $ " | Loss: " ++ show loss
+          liftIO $ print label 
+          (newParam, _) <- liftIO $ runStep model optimizer loss 1e-3
+          pure $ replaceParameters model newParam
+        done = pure
+        init = pure model
+
+main :: IO ()
+main = runSafeT $ do
+  init <- liftIO $ sample spec 
+  let irisTrain = (csvDataset @Iris "iris.data") { batchSize = 4 , shuffle = Just 500}
+
+  -- inputs <- makeListT @_ @_ @_ @[Iris] defaultDataloaderOpts irisTrain (Select $ yield ()) 
+  -- transformed <- pMap inputs Main.toTensors 2
+
+  foldM (\model epoch -> do
+            inputs <- makeListT @[Iris] defaultDataloaderOpts irisTrain (Select $ yield ()) 
+            transformed <- pMap inputs Main.toTensors 2
+            -- liftIO $ print epoch
+            trainLoop model optimizer $ enumerate transformed
+        ) init [1]
```

Comment:
I think dataloader has some waits when there are multiple epochs.

Is it possible to put the dataloader in an outer loop like this?
`replicate numEpoch dataload >-> transformer >-> training`
Does outer loop has any disadvantages?

Reply:
I'm not sure how we can avoid this, but yes there are some waits, in that the threads are killed and generated again on each epoch. I think the code you have written would still have this problem, I'll think more about this to see if it could be avoided but it shouldn't be too big of a bottleneck on real datasets.